### PR TITLE
Serialization in C#

### LINKED ISSUE #3216: Serialization in C#
### Proposed title of article
Serialization in C#

### Introduction paragraph (2-3 paragraphs):
The purpose of this article is to provide an overview of C# Serialization. Serialization is the process of converting an object instance into a data stream, and the state of the object instance is turned into a data stream so that it may be transferred over different networks and stored in a storage place.
This is an advantage of serialization because it allows you to send the converted data stream across different networks in a format that is cross-platform compatible and saves the converted stream data in a permanent or non-persistent object state so that a duplicate can be made at a later time.

### Key takeaways: 
1. Steps of C# Serialization Object
2. C# Serialization in action
3. Types of serialization that are supported by C#
4. C# program to demonstrate the concept of Serialization.

### References: 
N/A

### PR DESCRIPTION
**Pull request by Tonybishop11**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #3216 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Ericgacoki** [comment]
tab-crash

Hey @Tonybishop11, the author image you've provided is too large. Kindly crop it to around 600X600 PX and maintain it between 100KB and 1MB.

**Comment by Kayere** [comment]
PLagiarism check done ✔️

File: content/articles/serialization-in-csharp/index.md [hunk 1]
```diff
@@ -0,0 +1,169 @@
+### Introduction
+Serialization transforms a data structure or object state into a format that can be saved in a file, memory or sent over a computer network and then reconstructed. This is a benefit of Serialization in terms of transmitting the converted data.
+
+### Preliquisites
+- Have [visual studio 2019](https://visualstudio.microsoft.com/vs/) software installed. It is the compiler used to run Csharp programs.
```
Comment (Ericgacoki) [inline content/articles/serialization-in-csharp/index.md:5]: `Visual Studio 2019` is not a compiler. It's an IDE instead.

File: content/articles/serialization-in-csharp/index.md [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+- [Making a serializable object](#making-a-serializable-object)
+- [C# Serialization in action](#c#-serialization-in-action)
+- [Types of serialization in C#](Types-of-serialization-in-#c#)
+- [Conclusion](#conclusion)
+
+### Serialization applications
+Serialization allows developers to save an object's state and restore it later if needed. Serialization allows the following operations.
+
+- The item is delivered to the remote application through HTTP.
+- Domain to domain transfer - A JSON or XML data string can be used to transport an item over a firewall.
```
Comment (Ericgacoki) [inline content/articles/serialization-in-csharp/index.md:19]: Paraphrase the first list item to align with the intro sentence such that the two can be read without loss of flow.

File: content/articles/serialization-in-csharp/index.md [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+- Keeping user-specific or security-related data between apps.
+
+### Making a serializable object
+You'll need the following to serialize binary or XML data:
+- The serializable object
+- A stream will hold the serialized object.
+- It's a system.
+- Instance of Runtime.Serialization.Formatter
```
Comment (Ericgacoki) [inline content/articles/serialization-in-csharp/index.md:27]: How is `"It's a system"` applied as a requirement here? 🤔

File: content/articles/serialization-in-csharp/index.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+Create instances of the stream and the format to use, then use the Serialize function on the format. This call receives two strings as parameters: the stream and the object to serialize.
+
+### C# Serialization in action
+With apps, we must store data on durable or non-durable media for later retrieval. Serialization may help. Seriation is the process of storing an object instance as a byte stream in memory or a database. Serialization is needed to transmit an object over a network. Serializing in C# The app needs a serialization namespace. This is called serialization. In C#, utilize the [ Serializable ] property.
+
```
Comment (Ericgacoki) [inline content/articles/serialization-in-csharp/index.md:33]: You've already defined what Serialization is. You don't need to repeat that.

File: content/articles/serialization-in-csharp/index.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+{
+public int item;
+public string identity;
+[NonSerialized()] Public double cost; //We added the nonserialized symbol to make this class non-serializable.
+}
+```
+### Types of Serialization in C#
+
+1. **Binary Serialization**
+The System.Runtime.Serialization is a namespace that includes binary serialization classes. In computing, a namespace is a collection of symbols used to identify and refer to various things. A namespace ensures that all objects in a set have unique, easy-to-recognize names. Binary encoding provides a concise serialization for storage and socket-based network streams. With Binary Serialization, even read-only members are serialized, improving speed. This is the method for converting.NET objects into byte streams. During binary Serialization, all public, private, and read-only members are processed. Binary Serialization is used to quickly convert items to bytes. It utilizes System. Runtime. Serialization.Formatters.Binary reference.
```
Comment (Ericgacoki) [inline content/articles/serialization-in-csharp/index.md:58]: Split this into two.

File: content/articles/serialization-in-csharp/index.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            SoapFormatter formatter =new  SoapFormatter();
+            formatter.Serializa(fileStream, sample);
+        }
+    }
+}
+
+
+```
+
+From the above code, we created an instance of our sample class and called it erick nyaga, and then we serialized it using the value 34. After that we create a file stream to accept our output using the code `FileStream filestream =new Filestream(@"d:\sem\serialisation.dat", FileMode.Create);`. The line of the code after creating a file stream is used to serialize the object created. The final line of code serializes the objects to .data file format. The lines `using System.Runtime.Serialization;` and `using System.Runtime.Serialization.Formatters.Soap;` are the references for the code to do soap serialization.
```
Comment (Ericgacoki) [inline content/articles/serialization-in-csharp/index.md:152]: Quote the actual line(s) of code you're referring to and avoid using different tenses in the same paragraph/sentence.

File: content/authors/erick-kiragu/index.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+title: Erick Kiragu
+type: authors
+github: https://github.com/Tonybishop11
+---
+Student of data science with a passion for software development, artificial intelligence, and python scripting. I have proven my potential to expand technological knowledge by displaying exceptional abilities and capabilities. When I'm not coding, I enjoy playing video games and going to field competitions.
```
Comment (Ericgacoki) [inline content/authors/erick-kiragu/index.md:6]: This would sound better as, "`<name>` is a Data Science student..."

**Review by Ericgacoki — CHANGES_REQUESTED** [review]
Hey @Tonybishop11, kindly take a look at the following review-comments and ping me when you're done 😊

**Comment by Tonybishop11** [comment]
hello @Ericgacoki I have done the requested changes. Please check

**Review by Ericgacoki — APPROVED** [review]

**Comment by hectorkambow** [comment]
Thank you for your patience as our team has been working diligently to work through the recent increased volume of articles coming in. 

We are going through are ready articles in the order they were received and working on improving our review time as well. 

🚀

**Comment by lalith1403** [comment]
Taking this one 👍

**Comment by lalith1403** [comment]
> Plagiarism check done ✔️

What is the plagiarism percentage for this PR? Using another source, I got a score of 8%.

File: content/articles/serialization-in-csharp/index.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+### Making a serializable object
+You'll need the following to serialize binary or XML data:
+- The serializable object.
+- A stream to contain the serialized object.
+- A `System.Runtime.Serialization.Formatter` instance.
+
+Create instances of the stream and the format to use, then use the Serialize function on the format. This call receives two strings as parameters: the stream and the object to serialize.
+
+### C# Serialization in action
+With apps, we must store data on durable or non-durable media for later retrieval. Serialization may help. Serialization is needed to transmit an object over a network. Serializing in C#, the app needs a serialization namespace. In C#, utilize the [ Serializable] property.
```
Comment (lalith1403) [inline content/articles/serialization-in-csharp/index.md:53]: Why are you using the square brackets?

Comment (lalith1403) [reply content/articles/serialization-in-csharp/index.md:53]: If you are referring to the code, use backticks... \`\`.

File: content/articles/serialization-in-csharp/index.md [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+```
+
+### Types of Serialization in C#
+
+#### 1. Binary Serialization
+The `System.Runtime.Serialization` is a namespace that includes binary serialization classes. In computing, a namespace is a collection of symbols used to identify and refer to various things. A namespace ensures that all objects in a set have unique, easy-to-recognize names. The binary encoding provides a straightforward serialization for storage and socket-based network streams.
+
+With Binary serialization, even read-only members are serialized, improving speed. This is the method for converting .NET objects into byte streams. During binary serialization, all public, private, and read-only members are processed. Binary serialization is used to convert items to bytes quickly. It utilizes `System.Runtime. serialization.Formatters.Binary` reference.
+
+Below is a biliary Serialization example:
```
Comment (lalith1403) [inline content/articles/serialization-in-csharp/index.md:83]: check spelling.

File: content/articles/serialization-in-csharp/index.md [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+### Custom Serialization
+Custom serialization allows event serialization and de-serialization. Using an ISerializable interface may help. Custom serialization allows you to control which items are serialized and how. SerializableAttribute and ISerializable interfaces are required.
+
+Consider the following code, which demonstrates custom Serialization using the ISerializable interface:
+
+```csharp
+[Serializable] public class Push : ISerializable
+{
+public virtual void GetObjectData(SerializationInfo information, StreamingContext context)
+{
```
Comment (lalith1403) [inline content/articles/serialization-in-csharp/index.md:193]: Check the code. Brackets are not matched.

**Review by lalith1403 — CHANGES_REQUESTED** [review]
@Ericgacoki @Tonybishop11 The article has many errors. Please rectify all the errors.

**Comment by Tonybishop11** [comment]
Hello @Ericgacoki done correcting requested changes

**Comment by lalith1403** [comment]
Make sure all the errors are changed. I had only mentioned a few errors that I found on a glance.

**Comment by Ericgacoki** [comment]
Hey @lalith1403, I have made a few more necessary changes. Kindly take a look.
Apologies for the inconveniences caused! 🤝

**Review by lalith1403 — APPROVED** [review]

**Pull request merged**